u2f-tests/HID/u2f_util.cc

```diff
@@ -385,6 +371,40 @@ int U2Fob_apdu(struct U2Fob* device,
   return sw12;
 }
 
+int U2Fob_apdu(struct U2Fob* device,
+               uint8_t CLA, uint8_t INS, uint8_t P1, uint8_t P2,
+               const std::string& out,
+               std::string* in) {
+  uint8_t buf[4096];
+  size_t nc = out.size() ? (3 + out.size()) : 0;
+  uint8_t ne = nc ? 2 : 3;
+  size_t bufSize = 4 + nc + ne;
+
+  // Construct outgoing message.
+  memset(buf, 0xEE, sizeof(buf));
+  buf[0] = CLA;
+  buf[1] = INS;
+  buf[2] = P1;
+  buf[3] = P2;
+
+  uint8_t offs = 4;
+  if (nc) {
+    buf[offs++] = 0;  // extended length
+    buf[offs++] = (out.size() >> 8) & 255;
+    buf[offs++] = (out.size() & 255);
+    memcpy(buf + offs, out.data(), out.size());
+    offs += out.size();
+  }
```

Comment:
This was a deliberate choice, to encode Lc and Le separately, even if the branches look a little funny. I added comments that I hope will clarify why it's structured this way.